pinot-core/src/main/java/org/apache/pinot/core/query/reduce/HavingFilterHandler.java

```diff
@@ -0,0 +1,182 @@
+
+    AndRowMatcher(List<FilterContext> childFilters) {
+      int numChildren = childFilters.size();
+      _childMatchers = new RowMatcher[numChildren];
+      for (int i = 0; i < numChildren; i++) {
+        _childMatchers[i] = getRowMatcher(childFilters.get(i));
+      }
+    }
+
+    @Override
+    public boolean isMatch(Object[] row) {
+      for (RowMatcher childMatcher : _childMatchers) {
+        if (!childMatcher.isMatch(row)) {
+          return false;
+        }
+      }
+      return true;
+    }
+  }
+
+  /**
+   * OR filter matcher.
+   */
+  private class OrRowMatcher implements RowMatcher {
+    RowMatcher[] _childMatchers;
+
+    OrRowMatcher(List<FilterContext> childFilters) {
+      int numChildren = childFilters.size();
+      _childMatchers = new RowMatcher[numChildren];
+      for (int i = 0; i < numChildren; i++) {
+        _childMatchers[i] = getRowMatcher(childFilters.get(i));
+      }
+    }
+
+    @Override
+    public boolean isMatch(Object[] row) {
+      for (RowMatcher childMatcher : _childMatchers) {
+        if (childMatcher.isMatch(row)) {
+          return true;
+        }
+      }
+      return false;
+    }
+  }
+
+  /**
+   * Predicate matcher.
+   */
+  private class PredicateRowMatcher implements RowMatcher {
+    PostAggregationHandler.ValueExtractor _valueExtractor;
+    DataType _valueType;
+    PredicateEvaluator _predicateEvaluator;
+
+    PredicateRowMatcher(Predicate predicate) {
+      _valueExtractor = _postAggregationHandler.getValueExtractor(predicate.getLhs());
+      switch (_valueExtractor.getColumnDataType()) {
+        case INT:
+          _valueType = DataType.INT;
+          break;
+        case LONG:
+          _valueType = DataType.LONG;
+          break;
+        case FLOAT:
+          _valueType = DataType.FLOAT;
+          break;
+        case DOUBLE:
+          _valueType = DataType.DOUBLE;
+          break;
+        case STRING:
+          _valueType = DataType.STRING;
+          break;
+        case BYTES:
+          _valueType = DataType.BYTES;
+          break;
+        default:
+          throw new IllegalStateException();
+      }
+      _predicateEvaluator = PredicateEvaluatorProvider.getPredicateEvaluator(predicate, null, _valueType);
+    }
+
+    @Override
+    public boolean isMatch(Object[] row) {
+      Object value = _valueExtractor.extract(row);
+      switch (_valueType) {
+        case INT:
+          return _predicateEvaluator.applySV((int) value);
+        case LONG:
+          return _predicateEvaluator.applySV((long) value);
+        case FLOAT:
+          return _predicateEvaluator.applySV((float) value);
+        case DOUBLE:
+          return _predicateEvaluator.applySV((double) value);
+        case STRING:
+          return _predicateEvaluator.applySV((String) value);
+        case BYTES:
+          return _predicateEvaluator.applySV(((ByteArray) value).getBytes());
+        default:
+          throw new IllegalStateException();
```

Comment:
some message?

Reply:
It is not possible to run into this branch as the `_valueType` is set inside the class